### PR TITLE
Improve `data_model` validation

### PR DESCRIPTION
**Pull request by uvsmtid-gs**
*   Validate both ways: `search_props` <-> `index_props`.
*   Relax validation where possible (report stats instead).
*   Configure generic `GitRepoLoader` with `class_name`-s it is supposed to load.
*   Fine-tune search_props for `Git*` `data_envelope`-s.
*   Rename `MetadataDelegator` to `DataBackendDelegator`.
*   Move logging and tracking calls under `ProgressTracker`.
*   Report `report_collection_sizes`.
*   Spec `TODO_98_35_14_72.exclude_class_name_from_search_control.md`.

### CODE REVIEW

File: src/argrelay/custom_integ/ServiceLoader.py [hunk 1]
```diff
@@ -55,12 +55,13 @@ def list_data_models(
     ) -> list[DataModel]:
 
         return [
-            DataModel(
-                collection_name = ServiceEnvelopeClass.ClassCluster.name,
-                class_name = ServiceEnvelopeClass.ClassCluster.name,
-                # TODO: TODO_89_50_17_63: fine-tune list of `index_props`:
-                index_props = [e.name for e in ServicePropName],
-            ),
+            # TODO: Uncomment before merge:
```
Comment (uvsmtid-gs) [inline src/argrelay/custom_integ/ServiceLoader.py:58]: Do before merge.

File: src/argrelay/custom_integ/ServiceLoader.py [hunk 2]
```diff
@@ -425,6 +426,8 @@ def populate_TD_63_37_05_36_default(
                 ServicePropName.cluster_name.name: "prod-apac-downstream",
             },
         ])
+        # TODO: Remove before merge:
```
Comment (uvsmtid-gs) [inline src/argrelay/custom_integ/ServiceLoader.py:429]: Do before merge.

File: src/argrelay/custom_integ/ServiceLoader.py [hunk 3]
```diff
@@ -1220,6 +1223,8 @@ def populate_TD_76_09_29_31_overlapped(
                 ServicePropName.cluster_name.name: "dev-emea-downstream",
             },
         ])
+        # TODO: Remove before merge:
```
Comment (uvsmtid-gs) [inline src/argrelay/custom_integ/ServiceLoader.py:1226]: Do before merge.

File: src/argrelay/custom_integ/ServiceLoader.py [hunk 4]
```diff
@@ -1345,6 +1350,8 @@ def populate_TD_99_99_88_75_mutually_exclusive(
...
```
Comment (uvsmtid-gs) [inline src/argrelay/custom_integ/ServiceLoader.py:1353]: Do before merge.

File: src/argrelay/custom_integ/ServiceLoader.py [hunk 5]
```diff
@@ -1435,6 +1442,8 @@ def populate_TD_38_03_48_51_large_generated(
                     }
 
                     cluster_envelopes.append(generated_cluster)
+                    # TODO: Remove before merge:
```
Comment (uvsmtid-gs) [inline src/argrelay/custom_integ/ServiceLoader.py:1445]: Do before merge.

File: src/argrelay/custom_integ/ServiceLoader.py [hunk 6]
```diff
@@ -1521,6 +1530,8 @@ def populate_TD_43_24_76_58_single(
                 ServicePropName.cluster_name.name: "dev-emea-downstream",
             },
         ])
+        # TODO: Remove before merge:
```
Comment (uvsmtid-gs) [inline src/argrelay/custom_integ/ServiceLoader.py:1533]: Do before merge.

File: src/argrelay/custom_integ/ServiceLoader.py [hunk 7]
```diff
@@ -1617,6 +1628,8 @@ def populate_TD_39_25_11_76_missing_props(
             },
 
         ])
+        # TODO: Remove before merge:
```
Comment (uvsmtid-gs) [inline src/argrelay/custom_integ/ServiceLoader.py:1631]: Do before merge.

File: src/argrelay/relay_server/LocalServer.py [hunk 8]
```diff
@@ -415,6 +442,8 @@ def _post_validate_all_data_envelope_for_missing_index_prop_names(
             ):
                 progress_tracker.envelope_per_col_i += 1
                 progress_tracker.total_envelope_i += 1
+                # TODO: Each collection may contain multiple class_names -
+                #       assert it is according to one of the `data_model`.
```
Comment (uvsmtid-gs) [inline src/argrelay/relay_server/LocalServer.py:446]: Do before merge.

File: docs/feature_stories/FS_45_08_22_15.data_model_manipulation.md [hunk 9]
```diff
@@ -4,6 +4,7 @@ feature_title: data model manipulation
 feature_status: PARTIAL
 ---
 
+# TODO: This time:
```
Comment (uvsmtid-gs) [inline docs/feature_stories/FS_45_08_22_15.data_model_manipulation.md:7]: Do before merge.

File: docs/feature_stories/FS_74_69_61_79.get_set_data_envelope.md [hunk 10]
```diff
@@ -7,6 +7,7 @@ feature_status: PARTIAL
 TODO: Maybe rename "get set data envelope" to "get set envelope_container"?
       Actually, it is all about selecting a snapshot of `data_envelope`-s - maybe "get set snapshot"?
 
+# TODO: This time:
```
Comment (uvsmtid-gs) [inline docs/feature_stories/FS_74_69_61_79.get_set_data_envelope.md:10]: Do before merge.

File: src/argrelay/custom_integ/GitRepoLoader.py [hunk 11]
```diff
@@ -78,18 +78,21 @@ def list_data_models(
...
                 collection_name = GitRepoEnvelopeClass.ClassGitTag.name,
                 class_name = GitRepoEnvelopeClass.ClassGitTag.name,
+                # TODO: This time:
                 # TODO: TODO_89_50_17_63: fine-tune list of `index_props`:
                 index_props = [e.name for e in GitRepoPropName],
             ),
             DataModel(
                 collection_name = GitRepoEnvelopeClass.ClassGitCommit.name,
                 class_name = GitRepoEnvelopeClass.ClassGitCommit.name,
+                # TODO: This time:
```
Comment (uvsmtid-gs) [inline src/argrelay/custom_integ/GitRepoLoader.py:95]: Do before merge.

File: src/argrelay/custom_integ/GitRepoLoader.py [hunk 12]
```diff
@@ -78,18 +78,21 @@ def list_data_models(
...
                 collection_name = GitRepoEnvelopeClass.ClassGitRepo.name,
                 class_name = GitRepoEnvelopeClass.ClassGitRepo.name,
+                # TODO: This time:
                 # TODO: TODO_89_50_17_63: fine-tune list of `index_props`:
                 index_props = [e.name for e in GitRepoPropName],
             ),
             DataModel(
                 collection_name = GitRepoEnvelopeClass.ClassGitTag.name,
                 class_name = GitRepoEnvelopeClass.ClassGitTag.name,
+                # TODO: This time:
```
Comment (uvsmtid-gs) [inline src/argrelay/custom_integ/GitRepoLoader.py:88]: Do before merge.

File: src/argrelay/custom_integ/GitRepoLoader.py [hunk 13]
```diff
@@ -78,18 +78,21 @@ def list_data_models(
             DataModel(
                 collection_name = GitRepoEnvelopeClass.ClassGitRepo.name,
                 class_name = GitRepoEnvelopeClass.ClassGitRepo.name,
+                # TODO: This time:
```
Comment (uvsmtid-gs) [inline src/argrelay/custom_integ/GitRepoLoader.py:81]: Do before merge.

File: src/argrelay/custom_integ/ServiceLoader.py (hunk 1 above)
Comment (uvsmtid-gs) [reply src/argrelay/custom_integ/ServiceLoader.py]: Done.

File: src/argrelay/custom_integ/ServiceLoader.py (hunk 3 above)
Comment (uvsmtid-gs) [reply src/argrelay/custom_integ/ServiceLoader.py]: Done.

File: src/argrelay/custom_integ/ServiceLoader.py (hunk 5 above)
Comment (uvsmtid-gs) [reply src/argrelay/custom_integ/ServiceLoader.py]: Done.

File: src/argrelay/custom_integ/ServiceLoader.py (hunk 7 above)
Comment (uvsmtid-gs) [reply src/argrelay/custom_integ/ServiceLoader.py]: Done.

File: src/argrelay/custom_integ/GitRepoLoader.py (hunk 13 above)
Comment (uvsmtid-gs) [reply src/argrelay/custom_integ/GitRepoLoader.py]: Done.

File: src/argrelay/custom_integ/GitRepoLoader.py (hunk 12 above)
Comment (uvsmtid-gs) [reply src/argrelay/custom_integ/GitRepoLoader.py]: Done.

File: src/argrelay/custom_integ/GitRepoLoader.py (hunk 11 above)
Comment (uvsmtid-gs) [reply src/argrelay/custom_integ/GitRepoLoader.py]: Done.

File: docs/feature_stories/FS_45_08_22_15.data_model_manipulation.md (hunk 9 above)
Comment (uvsmtid-gs) [reply docs/feature_stories/FS_45_08_22_15.data_model_manipulation.md]: Done.

File: src/argrelay/custom_integ/ServiceLoader.py (hunk 6 above)
Comment (uvsmtid-gs) [reply src/argrelay/custom_integ/ServiceLoader.py]: Done.

File: src/argrelay/custom_integ/ServiceLoader.py (hunk 4 above)
Comment (uvsmtid-gs) [reply src/argrelay/custom_integ/ServiceLoader.py]: Done.

File: src/argrelay/custom_integ/ServiceLoader.py (hunk 2 above)
Comment (uvsmtid-gs) [reply src/argrelay/custom_integ/ServiceLoader.py]: Done.

File: src/argrelay/_version.py [hunk 14]
```diff
@@ -1,4 +1,4 @@
 # See `docs/dev_notes/version_format.md`:
 # Implements this:
 # https://stackoverflow.com/a/7071358/441652
-__version__ = "0.7.20"
+__version__ = "0.7.21.dev1"
```
Comment (uvsmtid) [inline src/argrelay/_version.py:4]: Finalize before merge.

**Review by uvsmtid — APPROVED** [review]

File: src/argrelay/_version.py (hunk 14 above)
Comment (uvsmtid-gs) [reply src/argrelay/_version.py]: Done.

**Review by uvsmtid — APPROVED** [review]

**Pull request merged**